HARQ feature for 5G LDPC

rabihchrabieh opened the pull request:
<!--
SPDX-FileCopyrightText: Copyright (c) 2021-2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
SPDX-License-Identifier: Apache-2.0
-->

## Description

Added HARQ support for 5G LDPC codes.
Created a Jupyter notebook describing the functionality and tests it.

Currently, the HARQ functionality does not support pruning (TODO for RV0, etc).
Graph mode is working but should be revisited.

All (or most) existing tests related to 5G LDPC pass.
No new unit test was created (TODO).

The additional API does not impact the existing API.

Fixed a few unrelated tests that were previously broken (cn_type changed to cn_update).


## Checklist

- [x] Detailed description
- [ ] Added references to issues and discussions
- [ ] Added / modified documentation as needed
- [x] Added / modified unit tests as needed
- [x] Passes all tests
- [x] Lint the code
- [x] Performed a self review
- [x] Ensure you Signed-off the commits. Required to accept contributions!
- ~~[ ] Co-authored with someone? Add Co-authored-by: user@domain and ensure they signed off their commits too.~~

rabihchrabieh committed:
Add HARQ support for 5G LDPC codes in Sionna

- Implemented HARQ mode for LDPC encoder/decoder with RV selection
- Added HARQ tutorial with AWGN BLER performance example
- Added unit tests

Signed-off-by: Rabih Chrabieh <7547980+rabihchrabieh@users.noreply.github.com>

rabihchrabieh commented:
Thank you, @SebastianCa, for the helpful feedback.

We can definitely move the "rv" to the function parameters — that makes sense and ensures compatibility with graph/XLA mode. We might also allow passing "n" optionally, in case the user wants to test combinations like (rv0, n0), (rv1, n1).

Regarding stacking all RVs inside the encoder: I think that adds unnecessary complexity. In most scenarios, we want to test RVs through time-varying channels, so the stacked output would need to be unstacked anyway before transmission — which somewhat defeats the purpose.

For the decoder, I agree it should remain stateless. A simple and flexible approach is to accumulate the received RVs outside the decoder. This keeps the decoder logic clean and allows users to apply any desired accumulation method or weighting strategy.

That said, we could optionally provide a utility method for accumulation within the class (but not tied to the decoder itself), to support convenience without introducing state.

Let me know if this approach works for you or if you'd prefer something more integrated.

I will extend the notebook into a tutorial with background on HARQ and RVs, and switch to Es/N₀ for plotting to better reflect varying code rates. I will also add unittests comparing the HARQ decoder to the classical (rv=0) case under equivalent redundancy.

Good catch: the change to test_mimo_flat_fading.py wasn’t intentional and must have slipped in during testing. I’ll remove it from the PR.

rabihchrabieh commented:
As an alternative, I’d like to propose a simpler structure when HARQ mode is activated:

- encode(u) returns the full codeword (n_cb); no rate matching inside the encoder.
- decode(llr) takes the full-length soft input (n_cb); no rate matching or accumulation inside the decoder.
- Rate matching (RV slicing) and accumulation are handled externally in the simulation.

This keeps encode and decode clean, stateless, and graph-compatible, and avoids repeated calls to encode(u) when testing multiple RVs — the full n_cb output is stored once and sliced as needed for retransmissions.

A harq_mode flag is provided as a function parameter to both encode and decode. This supports backward compatibility:

- When harq_mode=False, we follow the standard RV0 behavior: the first 2*Z bits are skipped, and pruning may be applied.
- When harq_mode=True, we start from bit 0, and pruning is disabled — suitable for HARQ retransmissions.

A first transmission can use harq_mode=False, while retransmissions typically use harq_mode=True.

Let me know if this direction would be acceptable.

SebastianCa commented:
Hi @rabihchrabieh,
my feeling is that an additional HARQ Block makes the API usage more complicated (from a user perspective); I would suggest keeping it inside of the encoder/decoder. 

Here are my thoughts:
If we pass a list of rvs to the encoder, the user still has the flexibility to directly generate all rv-versions during the first encoding.
Simple example:
1. ) `x = encode(u)` returns `[bs, n_ldpc]` bits
2. ) `x = encode(u, rv=["rv3",])` returns `[bs, 1, n_ldpc]` bits
3. ) `x = encode(u, rv=["rv0", "rv1"])`returns `[bs, 2, n_ldpc]` bits

As the decoder has no internal state, we must provide all rv versions anyhow. Here I would suggest to simply stack the inputs. In case a user wants to do advanced experiments, e.g., only decode "rv3", one could stack with zero llrs
```
x_rv3 = encoder(u, rv=["rv3",])
llr_3 = channel(x_rv3,...)
llr = tf.concat(tf.zeros((bs,2,n_ldpc), llr_rv3, axis=0)
x_hat = decoder(llr)
```
Otherwise, one could then just stack the different rv versions (e.g., from transmission over different instances of time-varying channels) and run the decoder once. For example
```
# initial transmission
x_rv1 = encoder(u, rv=["rv1",])
llr_1 = channel(x_rv1,...)
x_hat1 = decoder(llr_1)
# 1st HARQ
x_rv2 = encoder(u, rv=["rv2",])
llr_2 = channel(x_rv1,...)
llr = tf.concat((llr_1, llr_2))
x_hat2 = decoder(llr)
...
```

Or alternatively, do it as one-shot experiment (assuming the channel model supports this)
```
x = encoder(u, rv=["rv1","rv2"])
# possibly reshape for compatibility with advanced channel models
# x = tf.reshape(x, (-1,n_ldpc))
llr= channel(x,...)
# undo reshaping
# llr = tf.reshape(llr, (-1,2,n_ldpc))
x_hat = decoder(llr)
```

The overhead in the decoder should be fairly small as it boils down to gathering and adding llrs before decoding.

I would always disable pruning if `harq_mode` is active.

What do you think, does that make sense?

rabihchrabieh commented:
Hi @SebastianCa,

Thanks again for the detailed input.

Before moving forward, I wanted to clarify a few points to make sure I understand the proposed design fully — especially in the context of stacked RVs.

Suppose a user provides `[rv0, rv0, rv2, rv1]` to the encoder. Then:

1. **Encoder behavior**: I assume the encoder runs once, producing the full `n_ldpc` block, and then applies RV-specific zeroing to generate 4 stacked versions — each masking out the bits not selected by its corresponding RV.

2. **Channel interface**: The encoder’s output cannot be transmitted directly, since only `n` (not `n_ldpc`) bits are sent per RV. This implies a step is needed to reduce each RV’d version to its `n` active bits.  
   Is the plan to perform this reduction:
   - **Inside the encoder**, embedding rate matching directly into `encode()`?  
   - Or via a separate method (e.g., `get_rv_bits()`), to keep `encode()` clean while keeping rate matching within the same LDPC class?

3. **Decoder input realignment**: On the decoder side, if the user stacks received vectors (each of length `n`), they need to be realigned or zero-padded back to `n_ldpc` before decoding. This seems like another form of rate matching.  
   Should this happen:
   - Inside `decode()` itself, using an RV list?  
   - Or externally, through a method like `accumulate_rv_bits()` (of the LDPC class) that maps the `n` bits back into the full `n_ldpc` domain?  
   Also, just to confirm: the decoder should not assume a fixed RV order (e.g., always `[rv0, rv1, rv2, ...]`). Users may transmit any RVs in any order, or repeat RVs if needed.

Alternatively, is the idea to transmit all `n_ldpc` bits through the channel, even though only `n` are meaningful per RV — and then mask the untransmitted bits at the decoder (setting them to 0)? That would deviate from standard behavior and may complicate things if the user is constrained to transmit only `n` bits through various blocks (like interleaving, IFFT) and channel.

Just trying to get a clearer picture of how you see these components fitting together.

rabihchrabieh commented:
The issue is with displaying and printing the word EsNo on graphs and results tables. I had to make small changes.

SebastianCa commented:
Great - we'll review the code soon.

SebastianCa committed:
first code review iteration

modified docstrings
removed (uneded) public functions 
minor code changes

SebastianCa commented on src/sionna/phy/fec/ldpc/encoding.py:
```diff
@@ -39,23 +40,43 @@ class LDPC5GEncoder(Block):
+        If set to `True`, the encoder will allow coding rates below 1/3 and 1/5,
+        respectively for BG1 and BG2, consistent with 3GPP specifications.
+
     Input
```
Just a hint: the docstrings in the Sionna API are the class docstrings; the docstring of the call() are ignored (only for development). Thus, I've moved the docstrings from the call to here.

SebastianCa commented on src/sionna/phy/fec/ldpc/encoding.py:
```diff
@@ -39,23 +40,43 @@ class LDPC5GEncoder(Block):
+        any further initialization. This is useful to obtain the parameters
+        of the encoder without initializing the full block.
+
+    allow_low_rates: bool, (default `False`)
```
The reason we we don't allow low rates is that we don't support the 3GPP compliant rate-matching for such low rates, yet. 

My understanding of this flag is that it only disables the check and may produce a non-3GPP compliant code. This seems misleading to me and I would like to remove it.

rabihchrabieh replied:
I will remove it (but I need to see how to perform a particular regression test).

SebastianCa committed:
minor edits

SebastianCa committed:
adding compt

SebastianCa reviewed (commented):
I reviewed all files except the ipynb. This can be done once the code is ready

rabihchrabieh commented:
Hi @SebastianCa, thanks for your thorough review and good feedback.

I agree with most of your comments and suggestions but as I am currently on travel, it will take a bit of time to respond.

SebastianCa commented:
No worries - there is no deadline. We'll wait for your response.

rabihchrabieh commented on src/sionna/phy/utils/misc.py:
OK

rabihchrabieh commented on src/sionna/phy/fec/ldpc/decoding.py:
```diff
@@ -1402,6 +1443,44 @@ def __init__(self,
+            tensor_rv_dim = tf.shape(llr_ch)[-2]  # Second-to-last dimension
+            num_rv = len(rv_list)
+
+            tf.debugging.assert_equal(
```
Will change it.

rabihchrabieh commented on src/sionna/phy/utils/plotting.py:
OK

rabihchrabieh committed:
Restored misc.py and plotting.py to upstream version

Signed-off-by: Rabih Chrabieh <7547980+rabihchrabieh@users.noreply.github.com>

rabihchrabieh committed:
WIP: removed get_params and params_only

Signed-off-by: Rabih Chrabieh <7547980+rabihchrabieh@users.noreply.github.com>

rabihchrabieh committed:
WIP: removed an unnecessary XLA test and added more test combinations

Signed-off-by: Rabih Chrabieh <7547980+rabihchrabieh@users.noreply.github.com>

SebastianCa commented:
Hi @rabihchrabieh,

thanks for updating the PR; please expect some delays with the review due to traveling.

rabihchrabieh commented on src/sionna/phy/fec/ldpc/decoding.py:
```diff
@@ -1258,10 +1259,36 @@ class LDPC5GDecoder(LDPCBPDecoder):
+        redundancy) where successive LLR receptions are inserted in a circular
+        buffer and accumulated. The circular buffer is of size [batch_size, n_cb].
+
+    accumulator: `None` (default) | callable
```
Removed the accumulation method for simplicity

rabihchrabieh commented on src/sionna/phy/fec/ldpc/encoding.py:
```diff
@@ -590,13 +663,43 @@ def _encode_fast(self, s):
             raise ValueError("Last dimension must be of length k.")
 
-    def call(self, bits):
+    def call(self, bits, rv=None):
```
This likely complicates the code, and only partially solves the problem: if we have a different number of rv, then it needs to re-trace. If we set it to always max number of rv (how many retransmissions is that?) then it's a lot of overhead and complex to manage the masked rv. Let me know if you have a particular solution in mind.

rabihchrabieh commented on src/sionna/phy/fec/ldpc/decoding.py:
```diff
@@ -1424,53 +1503,78 @@ def build(self, input_shape, **kwargs):
-        nb_par_bits = (self.encoder.n_ldpc - k_filler
-                       - self.encoder.k - self._nb_pruned_nodes)
+        rv_starts = self.encoder.get_rv_starts()
+        start_pos = [rv_starts[rv_name] for rv_name in rv]
```
Combined into a single function

rabihchrabieh commented on src/sionna/phy/fec/ldpc/encoding.py:
```diff
@@ -91,13 +114,15 @@ def __init__(self,
         self._coderate = k / n
         self._check_input = True # check input for consistency (i.e., binary)
 
+        self._allow_low_rates = allow_low_rates # allow rates below 1/3 (BG1) and 1/5 (BG2)
```
The comment "allow actual code rates slightly larger than 948/1024" predates my changes.
My changes related to lower rate have been removed.

rabihchrabieh commented on src/sionna/phy/fec/ldpc/encoding.py:
```diff
@@ -590,13 +663,43 @@ def _encode_fast(self, s):
+                    f"Invalid RV name '{rv_name}'. Valid RV names are: {sorted(valid_rvs)}"
+                )
+
+    def _get_rv_starts(self) -> dict:
```
Done

rabihchrabieh commented on test/unit/fec/test_ldpc_5g_harq.py:
```diff
@@ -0,0 +1,417 @@
+    n_cb = ldpc_params.n_cb
+
+    # Reference encoder assumes no rate matching
+    encoder_ref = LDPC5GEncoder(k, n_cb, allow_low_rates=True)
```
I have removed the allow low rates flag. But then I need to find another solution for this regression test.

rabihchrabieh committed:
WIP: removed accumulator method, removed "allow low rates", merged rv_starts methods, added multi-batch test

Signed-off-by: Rabih Chrabieh <7547980+rabihchrabieh@users.noreply.github.com>

rabihchrabieh committed:
WIP: updated HARQ notebook

Signed-off-by: Rabih Chrabieh <7547980+rabihchrabieh@users.noreply.github.com>

rabihchrabieh committed:
WIP: minor edit

Signed-off-by: Rabih Chrabieh <7547980+rabihchrabieh@users.noreply.github.com>

rabihchrabieh commented on test/unit/fec/test_ldpc_5g_harq.py:
```diff
@@ -0,0 +1,417 @@
+    def encode_harq():
+        bits = source([batch_size, k])
+        x_ref = encoder_ref(bits)  # Shape: [batch_size, n_cb]
+        x_ref = tf.roll(x_ref, shift=starts["rv0"], axis=-1)  # Undo shift of RV0
```
I changed the logic a bit

rabihchrabieh committed:
Minor edit

Signed-off-by: Rabih Chrabieh <7547980+rabihchrabieh@users.noreply.github.com>

rabihchrabieh committed:
Corrected HARQ unit test

Signed-off-by: Rabih Chrabieh <7547980+rabihchrabieh@users.noreply.github.com>

rabihchrabieh commented:
Just pushed new commits — sorry for the delay! I've aligned with all your suggestions except one, which I've explained in the thread.

rabihchrabieh committed:
HARQ: fix RV starts, isolate HARQ path, add spec-conformance tests, extend tutorial

- Fix RV start positions to match TS 38.212 Table 5.4.2.1-2 floor formulas
- Isolate HARQ encoding/decoding path from legacy code to simplify review
- Add thorough RV selection unit test sweeping many (k, n) operating points
  against a spec-faithful reference with NULL-mask filler skipping
- Extend HARQ tutorial with Rayleigh fading and sudden-interference sections

Signed-off-by: Rabih Chrabieh <7547980+rabihchrabieh@users.noreply.github.com>

rabihchrabieh commented:
Hi,

I pushed a single commit that (1) fixes HARQ RV start positions per TS 38.212 Table 5.4.2.1-2, and (2) clearly isolates the HARQ encode/decode flow from the normal (non-HARQ) path, so it's easy to review the HARQ-specific changes and verify there are no side effects on existing behavior. I also added spec-conformance unit tests that compare the implementation against an independent reference with NULL-mask filler skipping across many (k, n) operating points, and extended the HARQ tutorial with Rayleigh fading and sudden-interference sections.

Best,
Rabih

SebastianCa reviewed (commented):
Hi @rabihchrabieh,

Sionna 2.0 has transitioned from TensorFlow to PyTorch. As a result, this PR (TensorFlow-based) cannot be merged into the current main branch.

That said, we highly value your contribution. We suggest that we migrate this implementation internally to the new PyTorch-based Sionna 2.0 framework. If ok with you, we would be happy to include it in the next release, and you will be credited as a contributor.

Please let us know if you’re okay with this approach. If so, we can close this PR.

Thanks again for the nice contribution, and apologies that this falls right in the middle of the transition.

rabihchrabieh commented:
Thank you, @SebastianCa.
That works well for me. Please feel free to convert it internally, and let me know if you need any assistance.

SebastianCa commented:
Closing this PR; will be included in the next release of Sionna.
Thanks again!
